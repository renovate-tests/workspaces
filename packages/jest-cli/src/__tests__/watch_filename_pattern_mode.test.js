/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 */

'use strict';

const chalk = require('chalk');
const {KEYS} = require('../constants');

const runJestMock = jest.fn();

let terminalWidth;

jest.mock('ansi-escapes', () => ({
  clearScreen: '[MOCK - clearScreen]',
  cursorDown: (count = 1) => `[MOCK - cursorDown(${count})]`,
  cursorHide: '[MOCK - cursorHide]',
  cursorRestorePosition: '[MOCK - cursorRestorePosition]',
  cursorSavePosition: '[MOCK - cursorSavePosition]',
  cursorShow: '[MOCK - cursorShow]',
  cursorTo: (x, y) => `[MOCK - cursorTo(${x}, ${y})]`,
}));

jest.mock(
  '../search_source',
  () =>
    class {
      constructor(context) {
        this._context = context;
      }

      findMatchingTests(pattern) {
        const paths = [
          './path/to/file1-test.js',
          './path/to/file2-test.js',
          './path/to/file3-test.js',
          './path/to/file4-test.js',
          './path/to/file5-test.js',
          './path/to/file6-test.js',
          './path/to/file7-test.js',
          './path/to/file8-test.js',
          './path/to/file9-test.js',
          './path/to/file10-test.js',
          './path/to/file11-test.js',
        ].filter(path => path.match(pattern));

        return {
          tests: paths.map(path => ({
            context: this._context,
            duration: null,
            path,
          })),
        };
      }
    },
);

jest.doMock('chalk', () => new chalk.constructor({enabled: false}));

jest.doMock('strip-ansi');
require('strip-ansi').mockImplementation(str => str);

jest.doMock(
  '../run_jest',
  () =>
    function() {
      const args = Array.from(arguments);
      const [{onComplete}] = args;
      runJestMock.apply(null, args);

      // Call the callback
      onComplete({snapshot: {}});

      return Promise.resolve();
    },
);

jest.doMock('../lib/terminal_utils', () => ({
  getTerminalWidth: () => terminalWidth,
}));

const watch = require('../watch');

const toHex = char => Number(char.charCodeAt(0)).toString(16);

const globalConfig = {watch: true};

afterEach(runJestMock.mockReset);

describe('Watch mode flows', () => {
  let pipe;
  let hasteMapInstances;
  let contexts;
  let stdin;

  beforeEach(() => {
    terminalWidth = 80;
    pipe = {write: jest.fn()};
    hasteMapInstances = [{on: () => {}}];
    contexts = [{config: {}}];
    stdin = new MockStdin();
  });

  it('Pressing "P" enters pattern mode', () => {
    contexts[0].config = {rootDir: ''};
    watch(globalConfig, contexts, pipe, hasteMapInstances, stdin);

    // Write a enter pattern mode
    stdin.emit(KEYS.P);
    expect(pipe.write).toBeCalledWith(' pattern â€º ');

    const assertPattern = hex => {
      pipe.write.mockReset();
      stdin.emit(hex);
      expect(pipe.write.mock.calls.join('\n')).toMatchSnapshot();
    };

    // Write a pattern
    ['p', '.', '*', '1', '0'].map(toHex).forEach(assertPattern);

    [KEYS.BACKSPACE, KEYS.BACKSPACE].forEach(assertPattern);

    ['3'].map(toHex).forEach(assertPattern);

    // Runs Jest again
    runJestMock.mockReset();
    stdin.emit(KEYS.ENTER);
    expect(runJestMock).toBeCalled();

    // globalConfig is updated with the current pattern
    expect(runJestMock.mock.calls[0][0].globalConfig).toEqual({
      onlyChanged: false,
      testNamePattern: '',
      testPathPattern: 'p.*3',
      watch: true,
      watchAll: false,
    });
  });

  it('can select a specific file name from the typeahead results', () => {
    const toUnixPathPattern = pathPattern => pathPattern.replace(/\\\\/g, '/');

    contexts[0].config = {rootDir: ''};
    watch(globalConfig, contexts, pipe, hasteMapInstances, stdin);

    // Write a enter pattern mode
    stdin.emit(KEYS.P);

    // Write a pattern
    ['p', '.', '*']
      .map(toHex)
      .concat([
        KEYS.ARROW_DOWN,
        KEYS.ARROW_DOWN,
        KEYS.ARROW_DOWN,
        KEYS.ARROW_UP,
      ])
      .forEach(key => stdin.emit(key));

    stdin.emit(KEYS.ENTER);

    expect(
      toUnixPathPattern(
        runJestMock.mock.calls[1][0].globalConfig.testPathPattern,
      ),
    ).toMatchSnapshot();
  });

  it('Results in pattern mode get truncated appropriately', () => {
    contexts[0].config = {rootDir: ''};
    watch(globalConfig, contexts, pipe, hasteMapInstances, stdin);

    stdin.emit(KEYS.P);

    [30, 25, 20].forEach(width => {
      terminalWidth = width;
      stdin.emit(KEYS.BACKSPACE);
      pipe.write.mockReset();
      stdin.emit(KEYS.A);
      expect(pipe.write.mock.calls.join('\n')).toMatchSnapshot();
    });
  });

  it('Shows the appropiate header when the filename filter is active', () => {
    contexts[0].config = {rootDir: ''};
    watch(globalConfig, contexts, pipe, hasteMapInstances, stdin);

    stdin.emit(KEYS.P);

    ['p', '.', '*', '1', '0']
      .map(toHex)
      .concat(KEYS.ENTER)
      .forEach(key => stdin.emit(key));

    pipe.write.mockReset();
    stdin.emit(KEYS.P);
    expect(pipe.write.mock.calls.join('\n')).toMatchSnapshot();

    ['p'].map(toHex).concat(KEYS.ENTER).forEach(key => stdin.emit(key));

    pipe.write.mockReset();
    stdin.emit(KEYS.P);
    expect(pipe.write.mock.calls.join('\n')).toMatchSnapshot();
  });

  it('Shows the appropiate header when the test name filter is active', () => {
    contexts[0].config = {rootDir: ''};
    watch(globalConfig, contexts, pipe, hasteMapInstances, stdin);

    stdin.emit(KEYS.T);

    ['t', 'e', 's', 't']
      .map(toHex)
      .concat(KEYS.ENTER)
      .forEach(key => stdin.emit(key));

    pipe.write.mockReset();
    stdin.emit(KEYS.T);
    expect(pipe.write.mock.calls.join('\n')).toMatchSnapshot();

    ['t'].map(toHex).concat(KEYS.ENTER).forEach(key => stdin.emit(key));

    pipe.write.mockReset();
    stdin.emit(KEYS.T);
    expect(pipe.write.mock.calls.join('\n')).toMatchSnapshot();
  });

  it('Shows the appropiate header when both filters are active', () => {
    contexts[0].config = {rootDir: ''};
    watch(globalConfig, contexts, pipe, hasteMapInstances, stdin);

    stdin.emit(KEYS.P);

    ['p', '.', '*', '1', '0']
      .map(toHex)
      .concat(KEYS.ENTER)
      .forEach(key => stdin.emit(key));

    stdin.emit(KEYS.T);
    ['t', 'e', 's', 't']
      .map(toHex)
      .concat(KEYS.ENTER)
      .forEach(key => stdin.emit(key));

    pipe.write.mockReset();
    stdin.emit(KEYS.T);
    expect(pipe.write.mock.calls.join('\n')).toMatchSnapshot();
  });

  it('Pressing "c" clears the filters', () => {
    contexts[0].config = {rootDir: ''};
    watch(globalConfig, contexts, pipe, hasteMapInstances, stdin);

    stdin.emit(KEYS.P);

    ['p', '.', '*', '1', '0']
      .map(toHex)
      .concat(KEYS.ENTER)
      .forEach(key => stdin.emit(key));

    stdin.emit(KEYS.T);
    ['t', 'e', 's', 't']
      .map(toHex)
      .concat(KEYS.ENTER)
      .forEach(key => stdin.emit(key));

    stdin.emit(KEYS.C);

    pipe.write.mockReset();
    stdin.emit(KEYS.P);
    expect(pipe.write.mock.calls.join('\n')).toMatchSnapshot();
  });
});

class MockStdin {
  constructor() {
    this._callbacks = [];
  }

  setRawMode() {}

  resume() {}

  setEncoding() {}

  on(evt, callback) {
    this._callbacks.push(callback);
  }

  emit(key) {
    this._callbacks.forEach(cb => cb(key));
  }
}
