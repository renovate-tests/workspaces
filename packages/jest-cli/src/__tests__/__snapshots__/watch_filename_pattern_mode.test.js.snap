// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Watch mode flows Pressing "P" enters pattern mode 1`] = `
"


 pattern › p
[MOCK - cursorSavePosition]


 Pattern matches 11 files

 › path/to/file1-test.js

 › path/to/file2-test.js

 › path/to/file3-test.js

 › path/to/file4-test.js

 › path/to/file5-test.js

 › path/to/file6-test.js

 › path/to/file7-test.js

 › path/to/file8-test.js

 › path/to/file9-test.js

 › path/to/file10-test.js

   ...and 1 more file
[MOCK - cursorTo(12, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "P" enters pattern mode 2`] = `
"


 pattern › p.
[MOCK - cursorSavePosition]


 Pattern matches 11 files

 › path/to/file1-test.js

 › path/to/file2-test.js

 › path/to/file3-test.js

 › path/to/file4-test.js

 › path/to/file5-test.js

 › path/to/file6-test.js

 › path/to/file7-test.js

 › path/to/file8-test.js

 › path/to/file9-test.js

 › path/to/file10-test.js

   ...and 1 more file
[MOCK - cursorTo(13, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "P" enters pattern mode 3`] = `
"


 pattern › p.*
[MOCK - cursorSavePosition]


 Pattern matches 11 files

 › path/to/file1-test.js

 › path/to/file2-test.js

 › path/to/file3-test.js

 › path/to/file4-test.js

 › path/to/file5-test.js

 › path/to/file6-test.js

 › path/to/file7-test.js

 › path/to/file8-test.js

 › path/to/file9-test.js

 › path/to/file10-test.js

   ...and 1 more file
[MOCK - cursorTo(14, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "P" enters pattern mode 4`] = `
"


 pattern › p.*1
[MOCK - cursorSavePosition]


 Pattern matches 3 files

 › path/to/file1-test.js

 › path/to/file10-test.js

 › path/to/file11-test.js
[MOCK - cursorTo(15, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "P" enters pattern mode 5`] = `
"


 pattern › p.*10
[MOCK - cursorSavePosition]


 Pattern matches 1 file

 › path/to/file10-test.js
[MOCK - cursorTo(16, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "P" enters pattern mode 6`] = `
"


 pattern › p.*1
[MOCK - cursorSavePosition]


 Pattern matches 3 files

 › path/to/file1-test.js

 › path/to/file10-test.js

 › path/to/file11-test.js
[MOCK - cursorTo(15, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "P" enters pattern mode 7`] = `
"


 pattern › p.*
[MOCK - cursorSavePosition]


 Pattern matches 11 files

 › path/to/file1-test.js

 › path/to/file2-test.js

 › path/to/file3-test.js

 › path/to/file4-test.js

 › path/to/file5-test.js

 › path/to/file6-test.js

 › path/to/file7-test.js

 › path/to/file8-test.js

 › path/to/file9-test.js

 › path/to/file10-test.js

   ...and 1 more file
[MOCK - cursorTo(14, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "P" enters pattern mode 8`] = `
"


 pattern › p.*3
[MOCK - cursorSavePosition]


 Pattern matches 1 file

 › path/to/file3-test.js
[MOCK - cursorTo(15, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Pressing "c" clears the filters 1`] = `
"[MOCK - cursorHide]
[MOCK - clearScreen]

Pattern Mode Usage
 › Press Esc to exit pattern mode.
 › Press Enter to apply pattern to all filenames.


[MOCK - cursorShow]



 pattern › 
[MOCK - cursorSavePosition]


 Start typing to filter by a filename regex pattern.
[MOCK - cursorTo(11, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Results in pattern mode get truncated appropriately 1`] = `
"


 pattern › a
[MOCK - cursorSavePosition]


 Pattern matches 11 files

 › path/to/file1-test.js

 › path/to/file2-test.js

 › path/to/file3-test.js

 › path/to/file4-test.js

 › path/to/file5-test.js

 › path/to/file6-test.js

 › path/to/file7-test.js

 › path/to/file8-test.js

 › path/to/file9-test.js

 › path/to/file10-test.js

   ...and 1 more file
[MOCK - cursorTo(12, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Results in pattern mode get truncated appropriately 2`] = `
"


 pattern › a
[MOCK - cursorSavePosition]


 Pattern matches 11 files

 › ...to/file1-test.js

 › ...to/file2-test.js

 › ...to/file3-test.js

 › ...to/file4-test.js

 › ...to/file5-test.js

 › ...to/file6-test.js

 › ...to/file7-test.js

 › ...to/file8-test.js

 › ...to/file9-test.js

 › ...o/file10-test.js

   ...and 1 more file
[MOCK - cursorTo(12, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Results in pattern mode get truncated appropriately 3`] = `
"


 pattern › a
[MOCK - cursorSavePosition]


 Pattern matches 11 files

 › ...st.js

 › ...st.js

 › ...st.js

 › ...st.js

 › ...st.js

 › ...st.js

 › ...st.js

 › ...st.js

 › ...st.js

 › ...t.js

   ...and 1 more file
[MOCK - cursorTo(12, 5)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Shows the appropiate header when both filters are active 1`] = `
"[MOCK - cursorHide]
[MOCK - clearScreen]

Active Filters: filename /p.*10/, test name /test/


Pattern Mode Usage
 › Press Esc to exit pattern mode.
 › Press Enter to apply pattern to all tests.


[MOCK - cursorShow]



 pattern › 
[MOCK - cursorSavePosition]


 Start typing to filter by a test name regex pattern.
[MOCK - cursorTo(11, 7)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Shows the appropiate header when the filename filter is active 1`] = `
"[MOCK - cursorHide]
[MOCK - clearScreen]

Active Filters: filename /p.*10/


Pattern Mode Usage
 › Press Esc to exit pattern mode.
 › Press Enter to apply pattern to all filenames.


[MOCK - cursorShow]



 pattern › 
[MOCK - cursorSavePosition]


 Start typing to filter by a filename regex pattern.
[MOCK - cursorTo(11, 7)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Shows the appropiate header when the filename filter is active 2`] = `
"[MOCK - cursorHide]
[MOCK - clearScreen]

Active Filters: filename /p/


Pattern Mode Usage
 › Press Esc to exit pattern mode.
 › Press Enter to apply pattern to all filenames.


[MOCK - cursorShow]



 pattern › 
[MOCK - cursorSavePosition]


 Start typing to filter by a filename regex pattern.
[MOCK - cursorTo(11, 7)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Shows the appropiate header when the test name filter is active 1`] = `
"[MOCK - cursorHide]
[MOCK - clearScreen]

Active Filters: test name /test/


Pattern Mode Usage
 › Press Esc to exit pattern mode.
 › Press Enter to apply pattern to all tests.


[MOCK - cursorShow]



 pattern › 
[MOCK - cursorSavePosition]


 Start typing to filter by a test name regex pattern.
[MOCK - cursorTo(11, 7)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows Shows the appropiate header when the test name filter is active 2`] = `
"[MOCK - cursorHide]
[MOCK - clearScreen]

Active Filters: test name /t/


Pattern Mode Usage
 › Press Esc to exit pattern mode.
 › Press Enter to apply pattern to all tests.


[MOCK - cursorShow]



 pattern › 
[MOCK - cursorSavePosition]


 Start typing to filter by a test name regex pattern.
[MOCK - cursorTo(11, 7)]
[MOCK - cursorRestorePosition]"
`;

exports[`Watch mode flows can select a specific file name from the typeahead results 1`] = `"path/to/file2-test.js"`;
